### ISSUE TITLE
Lines often don't draw on OS X 10.14.3

### ISSUE THREAD

**Issue by df-0** [opentoonz/opentoonz#2591]
### Issue Summary

When you move the cursor quickly and then put it down and draw, no line appears. This affects as many as 50% of the lines I (try to) draw, making the program more or less unusable. If I hold the cursor still for a second before I put it down to draw, it draws normally!

### Screenshot or Video Reference
http://sendvid.com/czxx8xm9
(it's running slower than normal because of the screen capture, but happens even at higher refresh rates)

### Steps to Reproduce
move the cursor very quickly and start drawing before the red circle has caught up. I was doing this on a toonz raster level, but it happens with toonz vector levels as well.

### Expected Results
Lines should draw, resulting in a beautiful picture.

### Actual Results
Many of the lines do not draw, resulting in a sad animator.

### System Information
* **OpenToonz Version**: 
1.3 (built Jan 30th 2019 10:31:02)
* **Operating System**: 
OS X 10.14.3
* **CPU**: 
2.8GHz Intel Core i7
* **Memory**: 
16GB 1600 MHz DDR3
* **Graphics Card**: 
NVidia GeForce GT 650RM 1024MB
* **Graphics Tablet**: 
Wacom Bamboo, Wacom Cintiq 13 Pro (I've used this on opentoonz windows and it works fine), also happens with trackpad when no tablet is connected

**Comment by jrhyme** [comment]
Brushes did not try to customize? ahaha) or pushing your pen?

**Comment by df-0** [comment]
No custom brushes, this is a clean install. Edited, yup I meant MacOS not OS X. Will try updating to 10.14.4 and see if there's a difference.

**Comment by jrhyme** [comment]
> No custom brushes, this is a clean install. And it happens even with pressure sensitivity turned off, or using the trackpad with no tablet connected. Edited, yup I meant MacOS not OS X. Will try updating to 10.14.4 and see if there's a difference.

Here the version of the program still plays the role, I use the last nightlies

**Comment by artisteacher** [comment]
I haven't run into this with 10.14.3 and a touchpad. Is there a difference without the Cintiq?

**Comment by jrhyme** [comment]
> I haven't run into this with 10.14.3 and a touchpad. Is there a difference without the Cintiq?

I do not know, I use Gaomon M106K.

**Comment by df-0** [comment]
> I haven't run into this with 10.14.3 and a touchpad. Is there a difference without the Cintiq?

See description - it happens with or without the cintiq - it even happens with no tablet, just drawing on the macbook trackpad. Additionally both wacoms work fine in other software eg krita, photoshop

**Comment by df-0** [comment]
> > No custom brushes, this is a clean install. And it happens even with pressure sensitivity turned off, or using the trackpad with no tablet connected. Edited, yup I meant MacOS not OS X. Will try updating to 10.14.4 and see if there's a difference.
> 
> Here the version of the program still plays the role, I use the last nightlies

I've now updated to MacOS 10.14.4, and the problem's still there. Updated bug description to reflect that. This is on the mainline build though - will try updating to the nightlies and see if it's still there

**Comment by jrhyme** [comment]
> > > No custom brushes, this is a clean install. And it happens even with pressure sensitivity turned off, or using the trackpad with no tablet connected. Edited, yup I meant MacOS not OS X. Will try updating to 10.14.4 and see if there's a difference.
> > 
> > 
> > Here the version of the program still plays the role, I use the last nightlies
> 
> I've now updated to MacOS 10.14.4, and the problem's still there. Updated bug description to reflect that. This is on the mainline build though - will try updating to the nightlies and see if it's still there

In Macos, a lot of bugs, I also wrote a report before your topic and I can find a lot more. In the version for windows, there are no such bugs.

**Comment by df-0** [comment]
> > No custom brushes, this is a clean install. And it happens even with pressure sensitivity turned off, or using the trackpad with no tablet connected. Edited, yup I meant MacOS not OS X. Will try updating to 10.14.4 and see if there's a difference.
> 
> Here the version of the program still plays the role, I use the last nightlies

I've now updated to the latest nightly build, but the bug's still there. I've updated the bug description to reflect that.

**Comment by jrhyme** [comment]
> > > No custom brushes, this is a clean install. And it happens even with pressure sensitivity turned off, or using the trackpad with no tablet connected. Edited, yup I meant MacOS not OS X. Will try updating to 10.14.4 and see if there's a difference.
> > 
> > 
> > Here the version of the program still plays the role, I use the last nightlies
> 
> I've now updated to the latest nightly build, but the bug's still there. I've updated the bug description to reflect that.

Ok, create as much as possible, maybe then developers will be more likely to edit the version under Macos)

**Comment by artisteacher** [comment]
We just need more developers who work with MacOS! Definitely add more issues as you find unreported bugs. Also, if you can give more information about any reported bugs, that’s helpful. The more specific information we can give, the more likely it is that the problem can be fixed. 

I finally was able to trigger this on my touchpad. It’s hard to trigger since it’s not natural to draw really swiftly on a touchpad. At first I thought it might be due to a high smooth value, but it can be triggered with smooth set to 0. It doesn’t seem to be related to touch controls. Zoom level and brush hardness both don’t seem to matter. Happens with or without mypaint brushes. Happens in both the comboviewer and the viewer. 

 I haven’t been able to trigger it so far with the paint brush tool nor on Raster levels.

**Comment by df-0** [comment]
> We just need more developers who work with MacOS! Definitely add more issues as you find unreported bugs. Also, if you can give more information about any reported bugs, that’s helpful. The more specific information we can give, the more likely it is that the problem can be fixed.
> 
> I finally was able to trigger this on my touchpad. It’s hard to trigger since it’s not natural to draw really swiftly on a touchpad. At first I thought it might be due to a high smooth value, but it can be triggered with smooth set to 0. It doesn’t seem to be related to touch controls. Zoom level and brush hardness both don’t seem to matter. Happens with or without mypaint brushes. Happens in both the comboviewer and the viewer.
> 
> I haven’t been able to trigger it so far with the paint brush tool nor on Raster levels.

yeah, I've definitely had it happen at zero smoothing. One thing I did find was quite effective at triggering it was (with tablet) drawing a stroke and lifting the pen up rapidly out of tracking range so the cursor remains in place at the end of the stroke, then starting a new stroke by suddenly putting the pen down. Once you've started a "stroke that doesn't draw" you can hold down the pen for as long as you want and nothing gets drawn- similarly once a stroke has successfully started, you can scribble as much as you want and it works fine.

**Comment by manongjohn** [comment]
There is definitely something going on with event processing here.  It is likely the tablet and/or mac is generating an odd series of events and the unexpected sequence OT is receiving is putting it in a strange state.  The issue may not necessarily be a common problem to all mac users, but may only affect certain macs under certain conditions.

I've seen stuff like this before with OT on my Windows machines.  On one machine with an external table, OT was acting normally, but on my other machine which has 2 tablet drivers on it (1 for a built in digitizer, 1 for an external tablet), I started having odd pen issues.  

With some debugging, I found both drivers were sending events even when the external tablet was disconnected. Depending on what those events are  and what order they came in caused problems while drawing.  I had to fix OT to ignore certain scenarios but I don't know what those scenarios are without debugging and seeing what events OT is receiving.

If your machine has multiple graphic tablet drivers on it from different tablets, does it help if you uninstalled one of them?  If you uninstalled all graphic tablet drivers, does the trackpad still have issues?

**Comment by df-0** [comment]
> There is definitely something going on with event processing here. It is likely the tablet and/or mac is generating an odd series of events and the unexpected sequence OT is receiving is putting it in a strange state. The issue may not necessarily be a common problem to all mac users, but may only affect certain macs under certain conditions.
> 
> I've seen stuff like this before with OT on my Windows machines. On one machine with an external table, OT was acting normally, but on my other machine which has 2 tablet drivers on it (1 for a built in digitizer, 1 for an external tablet), I started having odd pen issues.
> 
> With some debugging, I found both drivers were sending events even when the external tablet was disconnected. Depending on what those events are and what order they came in caused problems while drawing. I had to fix OT to ignore certain scenarios but I don't know what those scenarios are without debugging and seeing what events OT is receiving.
> 
> If your machine has multiple graphic tablet drivers on it from different tablets, does it help if you uninstalled one of them? If you uninstalled all graphic tablet drivers, does the trackpad still have issues?

Good thinking - I think both tablets share the same drivers (wacom 6.7.33-5, the current version) but will try uninstalling completely and see if the issue still occurs...

**Comment by df-0** [comment]
> If your machine has multiple graphic tablet drivers on it from different tablets, does it help if you uninstalled one of them? If you uninstalled all graphic tablet drivers, does the trackpad still have issues?
@manongjohn I've tested this, and updated the bug description with the test results. Sadly it didn't seem to help at all...

**Comment by kimandreasjohansson** [comment]
Just want to report that I have the exact same issue as stated above. 

Using a cintiq 16 (driver, 6.3.35-2, the new 16, not the pro version), tried to reinstall the drivers. No conflicting tablet drivers either. Like stated above you can reproduce this with a mouse or trackpad. Simply draw  something with the brush, quickly switch to eraser (command + e), erase a little, switch back (command + b) and draw again. Often the cursor will then simply get stuck or draw a pixel or two in the wrong location. To get it "unstuck" you have to lift the pen away from the screen and press again. (Extremely annoying, never had this issue in any other drawing app.)

This issue only happens in Mac os, when I boot camp into windows 10 on the same imac the issue is not there and drawing works just fine. 

Mac os 10.14.5
iMac (Retina 5K, 27-inch, 2017)
3,4 GHz Intel Core i5
16 GB 2400 MHz DDR4
Radeon Pro 570 4 GB

**Comment by manongjohn** [comment]
I think I may know what is going on here....

While debugging, Whenever I was drawing too fast, it would sometimes not draw and every time it wasn't drawing I was getting this message:

> force mouse press event

It's coming from here in `main.cpp`:
>
      if (eventType == "mac_generic_NSEvent") {
        if (IsLeftMouseDragged(message) && !leftButtonPressed) {
          std::cout << "force mouse press event" << std::endl;
          SendLeftMousePressEvent();
          return true;
        }
      }

Per Apple in this 2018 thread https://forums.developer.apple.com/thread/103992 the problem it seems is an accessibility change Apple started making in 10.14.  They modified `CGEventPost()` to require user permission to create events.  We use this function in `SendLeftMousePressEvent()` to generate mouse events when we encounter certain scenarios.   A few users have posted issues with CGEventPost() despite granting accessibility.  I believe a bug was opened.  It appears this might be what is causing this issue.

Not sure what the workaround is yet, or if a fix is even available. 

@shun-iwasawa If there is anything you may be able to do for this?

**Comment by kimandreasjohansson** [comment]
Aha, I realised I had not granted user permissions in security preferences, maybe I missed that after last update to 1.3. After I switched on permissions I experience this issue a lot less. Need more testing to be sure if it really fixed it though.

**Comment by artisteacher** [comment]
Interesting - I hadn't realized there was a permissions issue. I hope this can be resolved because there's no notification & most users won't think to check.

**Comment by kimandreasjohansson** [comment]
I have been trying this for a while now and the bug is definitely still there after permission switched on but i´m pretty sure it happens less.

**Comment by kimandreasjohansson** [comment]
This do not seem to be resolved in version 1.4 i’m afraid.

**Comment by piotrnikov** [comment]
Hello! Just to say that this problem still exist and is more than annoying. Quite the dealbreaker actually for a drawing program. What is weird is that all the other softwares work fine (Krita, Photoshop, ClipStudio, Blender,...).
Anyway to fix this?

**Comment by kimandreasjohansson** [comment]
Yes, there is no difference between 1.3 and 1.4 on this issue what I can tell. I can't say if the issues are definitely related but when the pen "gets stuck" or doesn't draw the intended line like described above, it aslo often draws a dot or line at random somewhere in the drawing. I´m gonna sit down again soon and try to document this better for someone who can actually code. (Not myself that is) If there is something particular a developer need us to test around this issue I'd be happy to help out.

**Comment by piotrnikov** [comment]
I think I have it solved but I'm not 100% sure. Please test the following.
This is what I've done (I'm on MacOs 10.15.4):
. I went to this folder:

<img width="848" alt="Screen_0001" src="https://user-images.githubusercontent.com/28474540/82115139-47c5f900-9761-11ea-9408-329a7054cdda.png">

. I opened System Preferences/Security & Privacy. I authorized/opened the padlock with my user password and then dragged and dropped the files from the Finder window to the following categories:
<img width="645" alt="Screen_0002" src="https://user-images.githubusercontent.com/28474540/82115207-b30fcb00-9761-11ea-86cd-c38c9ce3f462.png">
<img width="660" alt="Screen_0003" src="https://user-images.githubusercontent.com/28474540/82115211-b99e4280-9761-11ea-8a7b-768c9e4bc778.png">
<img width="668" alt="Screen_0004" src="https://user-images.githubusercontent.com/28474540/82115215-befb8d00-9761-11ea-9ce7-0e289c1021a3.png">

. I rebooted.
. Things appear so far to work as expected. Every line I draw is actually drawn!
Can anyone with the same problem try this?

**Comment by kimandreasjohansson** [comment]
I doubt (but not sure) that it is Wacom driver related since you can reproduce the "line getting stuck issue" by drawing with a mouse or trackpad, no tablet plugged in. I haven't actually tried without the Wacom drivers installed though, but have tried to reinstall the drivers without any luck fixing the issue. Also tested to turn on and off permissions, but it doesn't really seem to fix anything I´m afraid.

**I am totally sure that the bug only happens when drawing on a toonz raster layer or vector layer, not on a regular raster layer. Maybe this information can be helpful for a developer.**

Best.

**Comment by piotrnikov** [comment]
This bug does not happen on Tahoma for me!
I don't know what it means, but I'm glad it works :)

**Comment by kimandreasjohansson** [comment]
> This bug does not happen on Tahoma for me!
> I don't know what it means, but I'm glad it works :)

You are right, this issue does not seem to happen at all in Tacoma (Using 1.0.1). Great if that fix could feed back into OT 1.5.

**Comment by RodneyBaker** [comment]
This report could use a review as it has grown over 3 years stale.
If there is a solution in Tahoma2D for this it should be targeted for porting.
If the problem is resolved, recommend closing this report.

**Comment by andeon** [comment]
@RodneyBaker Since this is a very old report and there has been no feedback on the newest version of OpenToonz, which includes a customized Qt and other updates in C++ for example, the [nightly build](https://github.com/opentoonz/opentoonz/releases/tag/nightly), I recommend closing the issue.  

If the user still experiences problems after installing the new version, we could try to trace the updates in Tahoma2D, which reportedly does not have these issues, to see what fixed this problem.
